core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -121,9 +122,16 @@ object RequestChannel extends Logging {
       }
     }
 
-    def responseString(response: AbstractResponse): Option[String] = {
+    def requestLog: Option[JsonNode] = {
```

Comment:
We must use a `val` to store the `requestLog` in order to guarantee that the JSON representation is computed before the request is processed by the api layer. Otherwise, we still have a ProduceRequest without the data. The `val` may be private as well and I would put it right below `bodyAndSize` with a comment which explains why we do this.